workflow/controller/operator.go

```diff
@@ -1667,3 +1674,23 @@ func (woc *wfOperationCtx) checkAndCompress() error {
 
 	return nil
 }
+
+func (woc *wfOperationCtx) SubstituteParamsInVolumes(params map[string]string) error {
+	volumes := woc.wf.Spec.Volumes
+	volumesBytes, err := json.Marshal(volumes)
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+	fstTmpl := fasttemplate.New(string(volumesBytes), "{{", "}}")
+	newVolumesStr, err := common.Replace(fstTmpl, params, true)
+	if err != nil {
+		return err
+	}
+	var newVolumes []apiv1.Volume
+	err = json.Unmarshal([]byte(newVolumesStr), &newVolumes)
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+	woc.wf.Spec.Volumes = newVolumes
```

Comment:
I didn't catch this before, but we should almost never be modifying the user defined spec (with the exception of suspending/resuming). Doing so would break resubmit/retry. 

Instead of mutating the wf.Spec.Volumes, you can introduce a new private `[]Volume` field in workflowContext to store the substituted volumes. Then workflowpod.go would be modified to use this volume list instead of the woc.wf.Spec.Volumes.

It's fine if it means recalculate this on every reconciliation.

Reply:
It is true, we shouldn't overwrite it. But I think we don't since a new `woc` is created every time a `wf` is processed (`Create`/`Update`).

So, isn't `woc.wf` already a DeepCopy of `wf` [here](https://github.com/argoproj/argo/blob/master/workflow/controller/operator.go#L87)?

If I understand correctly, the `DeepCopy` of a Workflow DeepCopies every `wf` field ([code](https://github.com/argoproj/argo/blob/master/pkg/apis/workflow/v1alpha1/zz_generated.deepcopy.go#L906-L908)). So, it DeepCopies `wf.Spec`, which DeepCopies `wf.Spec.Volumes` ([code](https://github.com/argoproj/argo/blob/master/pkg/apis/workflow/v1alpha1/zz_generated.deepcopy.go#L974-L980)).

Reply:
Replying to [this](https://github.com/argoproj/argo/pull/1238#issuecomment-482663220):
> It is but we persist that copy at the end of reconciliation. So any modifications to spec will persist

Oh! You are right. I'm sorry, I had totally forgotten all about the `Update`s.
I updated the PR (also rebased on top of current `master`) :slightly_smiling_face: